capsules/src/led_matrix.rs

```diff
@@ -1,125 +1,118 @@
+//!         (3, 2),
+//!         (4, 2),
+//!         (0, 3),
+//!         (1, 3),
+//!         (2, 3),
+//!         (3, 3),
+//!         (4, 3),
+//!         (0, 4),
+//!         (1, 4),
+//!         (2, 4),
+//!         (3, 4),
+//!         (4, 4)
+//!     )),
 //! );
-//!
-//!
-//! Syscall Interface
-//! -----------------
-//!
-//! - Stability: 2 - Stable
-//!
-//! ### Command
-//!
-//! All LED operations are synchronous, so this capsule only uses the `command`
-//! syscall.
-//!
-//! #### `command_num`
-//!
-//! - `0`: Return the number of LEDs on this platform.
-//!   - `data`: Unused.
-//!   - Return: Number of LEDs.
-//! - `1`: Turn the LED on.
-//!   - `data`: The index of the LED. Starts at 0.
-//!   - Return: `Ok(())` if the LED index was valid, `INVAL` otherwise.
-//! - `2`: Turn the LED off.
-//!   - `data`: The index of the LED. Starts at 0.
-//!   - Return: `Ok(())` if the LED index was valid, `INVAL` otherwise.
-//! - `3`: Toggle the on/off state of the LED.
-//!   - `data`: The index of the LED. Starts at 0.
-//!   - Return: `Ok(())` if the LED index was valid, `INVAL` otherwise.
 
 use core::cell::Cell;
 
-use kernel::syscall::{CommandReturn, SyscallDriver};
 use kernel::utilities::cells::TakeCell;
-use kernel::{ErrorCode, ProcessId};
+use kernel::ErrorCode;
 
 use kernel::hil::gpio::{ActivationMode, Pin};
 use kernel::hil::led::Led;
 use kernel::hil::time::{Alarm, AlarmClient, ConvertTicks};
 
-/// Syscall driver number.
-use crate::driver;
-pub const DRIVER_NUM: usize = driver::NUM::Led as usize;
-
 /// Holds the array of LEDs and implements a `Driver` interface to
 /// control them.
-pub struct LedMatrixDriver<'a, L: Pin, A: Alarm<'a>> {
-    cols: &'a [&'a L],
-    rows: &'a [&'a L],
-    buffer: TakeCell<'a, [u8]>,
+pub struct LedMatrixDriver<
+    'a,
+    L: Pin,
+    A: Alarm<'a>,
+    const ROWS: usize,
+    const COLS: usize,
+    const LED_BITS: usize,
+> {
+    cols: &'a [&'a L; COLS],
+    rows: &'a [&'a L; ROWS],
+    buffer: TakeCell<'a, [u8; LED_BITS]>,
     alarm: &'a A,
     current_row: Cell<usize>,
     timing: u8,
     row_activation: ActivationMode,
     col_activation: ActivationMode,
 }
 
-impl<'a, L: Pin, A: Alarm<'a>> LedMatrixDriver<'a, L, A> {
+impl<'a, L: Pin, A: Alarm<'a>, const ROWS: usize, const COLS: usize, const LED_BITS: usize>
+    LedMatrixDriver<'a, L, A, ROWS, COLS, LED_BITS>
+{
     pub fn new(
-        cols: &'a [&'a L],
-        rows: &'a [&'a L],
-        buffer: &'a mut [u8],
+        cols: &'a [&'a L; COLS],
+        rows: &'a [&'a L; ROWS],
+        buffer: &'a mut [u8; LED_BITS],
         alarm: &'a A,
         col_activation: ActivationMode,
         row_activation: ActivationMode,
         refresh_rate: usize,
     ) -> Self {
         // Initialize all LEDs and turn them off
-        if (buffer.len() * 8) < cols.len() * rows.len() {
-            panic!("Matrix LED Driver: provided buffer is too small");
-        }
+        assert!(ROWS * COLS / 8 + 1 == LED_BITS);
```

Comment:
Document this assertion, explain that this is used until const generics https://github.com/rust-lang/rust/issues/76560 will be stable.